### DISCUSSION TITLE
When?....

### DISCUSSION THREAD

**Discussion by art926** [muerrilla/stable-diffusion-Latentshop#1] [General]
Hey Muerrilla!
Any plans for release? You showed me some teaser a while ago (https://user-images.githubusercontent.com/48160881/229326346-943141aa-f84d-47d9-b4df-5544cdd72b56.png), but haven't updated since then...

**Comment by muerrilla** [comment]
Oh man! I'm so sorry. I got caught up in my day job, and then the webui updates broke everything, so I procrastinated it into oblivion. Anyway, didn't know I had already set up the repo, looks like we're half-way there already! I'll tell you what. I can do some code cleanup and upload it right away, but you'll have to figure out everything by yourself until I do the docs. You up for that?

**Reply by art926** [reply]
Oh course! Can you explain theoretically how your approach work? Do you scale down the latent values somehow? Using what formulas?

**Comment by muerrilla** [comment]
Sure thing. Beware though, as this is more alchemy than science.

Here's a generation with the 4 channels of the latent:
<img src="https://github.com/muerrilla/Latentshop/assets/48160881/809ad6ae-883a-4d63-9687-7336e3e8e972" width="128">
![latent0 jpeg-0000](https://github.com/muerrilla/Latentshop/assets/48160881/52e361d1-cf29-43a9-a005-94de11d9df42)![latent1 jpeg-0000](https://github.com/muerrilla/Latentshop/assets/48160881/36a16570-dd3f-496b-ad33-475e84147fb9)![latent2 jpeg-0000](https://github.com/muerrilla/Latentshop/assets/48160881/6067ea00-10fe-4984-8bb4-4e1cab4d3161)![latent3 jpeg-0000](https://github.com/muerrilla/Latentshop/assets/48160881/b718b88b-28fb-477d-a62f-c0c3cf4525eb)

So... the latents are basically 32-bit (or is it 16? anyway, good dynamic range) 4-channeled images for all we care, and our problem is too much contrast. So why not just adjust the contrast of the latents before the VAE decodes them into 8-bit 3-channeled images? 

We know the latent values are almost in the range [-3,3] (which expands during the sampling process with higher CFG values). That makes 0 almost grey. So by multiplying the latent values in a number between 0 and 1 we could bring them closer to grey (lower contrast), or better yet, normalize the values to [0,1] range, then raise them to a power between 0 and 1 or above 1, to increase and decrease the contrast with a sigmoid curve (keeping the blacks and whites in place, and just shifting grey values towards black or white). 

There remain a few problems:

1- A latent full of zeroes does not really decode to grey, but the ugly beige/yellowish grey commonly referred to as "latent grey", which people say is the average of the latent space or whatever. Don't know, don't care. We deal with this by first creating a tiny grey image. Then we encode that into a "grey latent". Now we can subtract the grey latent from our latent before applying the adjustment (centering the latent on real grey) and add it back into it again after we've done the adjustment. This way as we push the values towards zero, they tend to become grey, not beige.

<img src="https://github.com/muerrilla/Latentshop/assets/48160881/0e43cbb0-b49b-4850-822e-4230fa269db7" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/b10bc5d9-d2a9-433e-970a-834b6afdcfae" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/a334c499-2b03-4092-8f9b-c4b4ac4a3379" width="128">
<img src="https://github.com/muerrilla/Latentshop/assets/48160881/0e43cbb0-b49b-4850-822e-4230fa269db7" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/b5c3f3b0-3a98-472b-ba21-faca87629655" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/8951d691-4d89-481d-a80e-b6dc71f99d72" width="128">
Left to right: No adjustment, some adjustment, a lot of adjustment, without grey correction (above) and with grey correction (below)

(Also, now that we have a grey latent, we can play with another way of reducing the contrast: interpolating between our latent and the grey latent, which is basically the same as the multiplication method described above, but with grey correction. so I provide the option for both. Think of it as photoshop's modern vs. legacy contrast modes.)

2- Turns out the latents are not really normal images, and applying image adjustments to them like this is actually pretty destructive! But the solution to that is easy! Instead of applying the adjustment when the sampling is done (just before decoding the latent) we can schedule it to be applied earlier on in the sampling process. This way the sampler will have time to "heal" the destructive effects of our adjustment! 
<img src="https://github.com/muerrilla/Latentshop/assets/48160881/1eab761f-1c61-415f-8ed9-c7335a54b1e5" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/44336430-f064-42b6-b9cf-6494911e8666" width="128">
Left: Adjustment applied after sampling, Right: Adjustment applied during sampling (note how the adjustment is more intense but the image is cleaner, though a bit blurry. We deal with that later.)

Now this is where the fun begins. By adjusting the mode and amount of adjustment, the scheduling of the adjustment (it makes a huge difference "when" in the sampling process we manipulate the latent), or changing the color of our grey latent, we're able to produce many different effects (brightness/contrast, color grading, saturation, etc.). But the most fun is changing how much the adjustment is applied to each channel of the latent. Turns out channel 0 affects luminosity, and the 3 other channels affect color. So you can, for example, lower the adjustment on the 3 color channels to prevent the color wash out effect when lowering the contrast. Or you can play around with the channel mix and it will do crazy filmic color processing effects!
<img src="https://github.com/muerrilla/Latentshop/assets/48160881/ec354f82-c3cf-4739-bd2e-9482d3fdd62d" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/a4f640d1-bd33-4621-8604-a90ee479c7da" width="128"><img src="https://github.com/muerrilla/Latentshop/assets/48160881/ef15f794-8ff6-4abb-9ed2-f2545b832f6e" width="128">

Your problem (controlling the destructive force of the amazing high CFG values!) falls within the category of setting the right schedule for and applying the right amount of "decontrasting". You'll have to do a bit of trial and error for every generation, but it usually won't take that long once you've got the hang of the method.

Anyway, hope that gave you some ideas. Codes is almost there. Will upload tonight or tomorrow.

**Reply by art926** [reply]
This is so cool, can’t wait to try it!!
Could you explain key differences between your approach, and what they use in the existing solution (sd-dynamic-thresholding) when used with the const “mimic” value?

**Reply by muerrilla** [reply]
I'll try. I haven't looked properly into the Dynamic Thresholding extension though, and have no idea what a mimic is. I know about the CFG rescale method, which I think is in the same category.

Basically, they are totally different animals. DT and the like are methods designed to address the specific problem of high CFG values. Mine is a general image adjustment method which can be applied to any condition. Doesn't matter if you're generating with CFG scale of 2 or 20, you can apply my method regardless to adjust the colors. Also, controlling high CFG through decontrasting is just a small part of what latentshop can do.

DT works by preventing the sampler from shooting the latent values into the no-no zone. So basically it changes the path of the diffusion, and you'll have changes not just in the color, but the overall structure and details of the generation. Also, it doesn't do the grey correction, so it pushes the image towards beige instead of neutral grey.

Take this CFG 30 generation as an example:

![colorful portrait of psychedelic medusa by steve mccurry-oGz-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/beabb4ff-b3e6-40b3-a2de-60dc22136718)

Applying enough CFG rescaling (and I guess almost the same goes for dynamic thresholding) to it to bring the colors within an acceptable range, will produce a totally different image:

![colorful portrait of psychedelic medusa by steve mccurry-0ch-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/02ad4375-7715-4c09-ac17-86f750c9986d)

It will help if we schedule the CFG rescaling to skip the first few sampling steps, so it doesn't affect the overall structure of the image:

![colorful portrait of psychedelic medusa by steve mccurry-G8x-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/4e989381-4b58-4a3c-905f-38aec7194712)

With my method you can apply just the right amount of adjustment at the very latest steps of sampling so it doesn't change the generation, and I'm gonna go out on a limb and claim that it generally produces higher fidelity:

![colorful portrait of psychedelic medusa by steve mccurry-uty-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/903c32d8-b6a9-472c-9ed1-0315b91aed71)

But more importantly, since it's a general purpose image adjustment method, you can do other things with it too:

![colorful portrait of psychedelic medusa by steve mccurry-sVe-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/517e25f9-9b50-442e-b810-345ea30de9b4)
![colorful portrait of psychedelic medusa by steve mccurry-Y1m-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/6fdce5d4-d111-4365-b1d6-d6ce5dbdef88)
![colorful portrait of psychedelic medusa by steve mccurry-ue4-0](https://github.com/muerrilla/stable-diffusion-Latentshop/assets/48160881/9416e2ab-4adf-4949-88a9-a71f6e7a95cd)

That being said, I quite often use both CFG Rescale and Latentshop together.

**Comment by muerrilla** [comment]
There you go. Latentshop is up!

**Reply by art926** [reply]
Thank you so much! I'm playing with it and it seems very cool!
Unfortunately, as you've mentioned, there are no perfect parameters seems like. It all depends so much on the prompt, step count and a sampler... But it's fun to play and possible to achieve really good results!

**Comment by ride5k** [comment]
still struggling to decipher what the controls do... would love a more fleshed out readme!
